core/src/main/scala/kafka/log/Log.scala

```diff
@@ -870,6 +870,10 @@ class Log(@volatile var dir: File,
   }
 
   private def loadProducerState(lastOffset: Long, reloadFromCleanShutdown: Boolean): Unit = lock synchronized {
+    // Pass the set of loaded segment base offsets to diff against the set of producer state snapshot
+    // files. Any orphaned files should be cleaned up.
+    val keepOffsets = logSegments.map(_.baseOffset)
+    producerStateManager.cleanupOrphanSnapshotFiles(keepOffsets)
```

Comment:
This should be moved to `loadLog()` since it only needs to run at startup. Also, the set of offsets to delete should exclude the log end offset, since that is snapshotted out during shutdown.